src/rand.rs

```diff
@@ -237,29 +233,19 @@ mod urandom {
 #[cfg(all(target_os = "linux", feature = "dev_urandom_fallback"))]
 mod sysrand_or_urandom {
     use crate::error;
+    use spin::Once;
 
-    enum Mechanism {
-        Sysrand,
-        DevURandom,
+    fn sysrand_supported() -> bool {
+        let mut dummy = [0u8; 1];
+        super::sysrand_chunk::chunk(&mut dummy[..]).is_ok()
     }
 
     pub fn fill(dest: &mut [u8]) -> Result<(), error::Unspecified> {
-        use lazy_static::lazy_static;
-
-        lazy_static! {
-            static ref MECHANISM: Mechanism = {
-                let mut dummy = [0u8; 1];
-                if super::sysrand_chunk::chunk(&mut dummy[..]).is_err() {
-                    Mechanism::DevURandom
-                } else {
-                    Mechanism::Sysrand
-                }
-            };
-        }
-
-        match *MECHANISM {
-            Mechanism::Sysrand => super::sysrand::fill(dest),
-            Mechanism::DevURandom => super::urandom::fill(dest),
+        static SYSRAND_SUPPORTED: Once<bool> = Once::new();
+        if *SYSRAND_SUPPORTED.call_once(sysrand_supported) {
+            super::sysrand::fill(dest)
```

Comment:
On Linux, at least, `getrandom()` can block a long time waiting for the kernel entropy pool to be initialized. It seems wrong to have all threads spinning while we wait for the syscall to complete. In theory all these spinning userspace threads could take away CPU time from the kernel and/or other processes that would be useful for initializing the entropy pool and allowing the kernel to return to userspace.

How about, instead, we keep track of whether `getrandom()` is supported using a tri-state atomic `usize`:

if 0, call `getrandom()` to see if it is available, atomically set the flag to 1 or 2 depending on the result, and continue as follows.
If 1, use `sysrand::fill()` to service the request.
if 2, use `urandom::fill()` to service the request.

The effect of this would be to push the mutex down to the kernel, out of this userspace process. Presumably the kernel isn't going to use spin locks in such a way that would cause it to start itself of CPU time needed to initialize the entropy pool.

WDYT?